pinot-core/src/main/java/org/apache/pinot/core/query/request/context/QueryContext.java

```diff
@@ -658,9 +665,26 @@ public QueryContext build() {
       queryContext._isUnsafeTrim =
           !queryContext.isSameOrderAndGroupByColumns(queryContext) || queryContext.getHavingFilter() != null;
 
+      // Pre-calculate group by trim size
+      queryContext._groupByTrimSize = getGroupByTrimSize(queryContext);
+
       return queryContext;
     }
 
+    private int getGroupByTrimSize(QueryContext queryContext) {
```

Comment:
This applies to segment level trim only (not server level trim). Let's rename the variable to reflect that (e.g. `effectiveSegmentGroupTrimSize`)